mypy/checker.py

```diff
@@ -3286,7 +3317,10 @@ def check_assignment(
                             lvalue_type = make_optional_type(lvalue_type)
                             self.set_inferred_type(lvalue.node, lvalue, lvalue_type)
 
-                    rvalue_type = self.check_simple_assignment(lvalue_type, rvalue, context=rvalue)
+                    rvalue_type, lvalue_type = self.check_simple_assignment(
+                        lvalue_type, rvalue, context=rvalue, inferred=inferred, lvalue=lvalue
+                    )
+                    inferred = None
```

Comment:
FWIW I don't understand why inferred to `None` is needed here. A more general comment: this method is now way too long, maybe it is a good time to split it into meaningfully named parts?

Reply:
It's needed so that we won't override the inferred type later in the function. I will add a comment. I agree that this method is hard to follow. I'll look into refactoring it in a follow-up PR.